File: plugins/iosystem/tutorial/ios_1_system.cpp
/** @file ios_1_system.cpp  

Tutorial, I/O systems


@ingroup Tutorials 

@include ios_1_system.cpp

*/

#include "libfaudes.h"

// make the faudes namespace available to our program
using namespace faudes;



/////////////////
// main program
/////////////////

int main() {

  ////////////////////////////////////////////
  // Constructors 
  ////////////////////////////////////////////

  // construct from file
  IoSystem sys1("data/ios_system1.gen");

  // copy construct
  IoSystem sys2(sys1);

  // assign 
  IoSystem sys3 = sys2;

  // inspect result on console
  std::cout << "################################\n";
  std::cout << "# tutorial, io system 1 \n";
  sys3.Write();
  std::cout << "################################\n";

  // graph output requires,  "dot" binary in $PATH
  try {
    sys1.GraphWrite("tmp_ios_system1.svg");
  } catch(faudes::Exception& exception) {
    std::cout << "warning: cannot execute graphviz' dot. " << std::endl;
  } 

  // output for doku
  sys1.Write("tmp_ios_system1.gen");


  ///////////////////////////////////////////////
  // Inspect/Edit I/O attributes
  ///////////////////////////////////////////////

  // events, result on console
  std::cout << "################################\n";
  std::cout << "# input events, output events\n";
  sys3.InputEvents().Write();
  sys3.OutputEvents().Write();
  std::cout << "################################\n";

  // maniputlate some state attributes
  // (to configure set state, see also IoSystem())
  sys3.SetOutputState("s1");
  sys3.SetInputState("s3");
  sys3.ClrOutputState("s2");

  // swap event status all together
  EventSet inevs = sys3.InputEvents();
  EventSet outevs = sys3.OutputEvents();
  sys3.ClrOutputEvent(outevs);
  sys3.SetInputEvent(outevs);
  sys3.ClrInputEvent(inevs);
  sys3.SetOutputEvent(inevs);

  // inspect result on console
  std::cout << "################################\n";
  std::cout << "# tutorial, manipulated io system 1 \n";
  sys3.Write();
  std::cout << "################################\n";
       
  ///////////////////////////////////////////////
  // Test I/O structure
  ///////////////////////////////////////////////

  // the test will set the state attributes accordingly
  bool isio=IsIoSystem(sys1);

  // sys1 from file will pass the test
  if(!isio) std::cout << "#### test FAILED (test case error) \n";

  // report
  std::cout << "################################\n";
  std::cout << "# tutorial, sys1 test io properties \n";
  sys1.InputStates().Write();
  sys1.OutputStates().Write();
  std::cout << "################################\n";

  // however, the input is not free
  bool isfree=IsInputLocallyFree(sys1);
  if(isfree) std::cout << "#### test PASSED (test case error) \n";
  IoFreeInput(sys1);

  // report
  std::cout << "################################\n";
  std::cout << "# tutorial, sys1 with free input\n";
  sys1.Write();
  std::cout << "################################\n";

  // output for doku
  sys1.Write("tmp_ios_system1f.gen");

  // however, the input is still not omega-free
  StateSet qerr;
  bool iswfree=IsInputOmegaFree(sys1,qerr);
  if(iswfree) std::cout << "#### test PASSED (test case error) \n";

  // report
  std::cout << "################################\n";
  std::cout << "# tutorial, sys1 error states\n";
  sys1.WriteStateSet(qerr);
  std::cout << "################################\n";

  // Record testcase
  FAUDES_TEST_DUMP("sys1", sys1);
  FAUDES_TEST_DUMP("isio", isio);
  FAUDES_TEST_DUMP("isfree", isfree);
  FAUDES_TEST_DUMP("iswfree", iswfree);
  FAUDES_TEST_DUMP("qerr", qerr);

  return 0 ;

}

File: plugins/iosystem/tutorial/ios_1_system.lua
-- Test/demonstrate I/O systems

-- Optionally, turn on faudes debugging messages
faudes.Verbosity(0)

-- ------------------------------------------
-- IoSystem basic maintenance
-- ------------------------------------------

-- Announce
print("################# IoSystem basic maintenance")

-- Construct from file
sys = faudes.IoSystem("data/ios_system1.gen")

-- Report inputs and outputs
print("Input events: ", sys:InputEvents());
print("Output events: ", sys:OutputEvents());
sys:SWrite()

-- Graphical output
sys:GraphWrite("tmp_ios_system1.png")

-- Have graphical of tank system for html docu
tank = faudes.IoSystem("data/ios_tank.gen")
tank:GraphWrite("tmp_ios_tank.png")


-- Record test case
FAUDES_TEST_DUMP("system1",sys)



-- ------------------------------------------
-- Access individual attributes
-- ------------------------------------------

print("################# IoSystm access attributes")

-- Test
if sys:InputEvent("u1") then
  print("'u1' is an input event of ", sys:Name(),  "(expected)")
end
if sys:OutputEvent("u1") then
  print("'u1' is an output event of ", sys:Name(), "(test case error!)")
end

-- Manipulate 
sys:SetOutputEvent("u1")
sys:SetErrorState("s1")
sys:ClrOutputEvent("y1")

-- Report
print("Input events: ", sys:InputEvents());
print("Output events: ", sys:OutputEvents());
print("Error states: ", sys:StateSetToString(sys:ErrorStates()));

-- fix system1
sys = faudes.IoSystem("data/ios_system1.gen")

-- ------------------------------------------
-- Testing I/O properties
-- ------------------------------------------

print("################# Test I/O properties: IOSystem1")

-- basic I/O
if faudes.IsIoSystem(sys) then
  print("Basic I/O properties: ok")
else
  print("Basic I/O properties: failed");
  print("Error states: ", sys:StateSetToString(sys:ErrorStates()));
end

-- input locally free
if faudes.IsInputLocallyFree(sys) then
  print("Input locally free: ok");
else
  print("Input locally free: failed");
  print("Error states: ", sys:StateSetToString(sys:ErrorStates()));
end

-- input omega freey
if faudes.IsInputOmegaFree(sys) then
  print("Input omega-free: ok");
else
  print("Input omega-free: failed");
  print("Error states: ", sys:StateSetToString(sys:ErrorStates()));
end

-- test XML token IO
sys:XWrite('tmp_ios_sys.ftx')
sys:Read('tmp_ios_sys.ftx')

-- record test case
FAUDES_TEST_DUMP("xml io",sys)

print("################# Test I/O properties: IOTank")

-- have a state partition
faudes.IoStatePartition(tank)
print("Input states: ", tank:StateSetToString(tank:InputStates()));
print("Output states: ", tank:StateSetToString(tank:OutputStates()));

-- basic I/0
if faudes.IsIoSystem(tank) then
  print("Basic I/O properties: ok (expected)")
else
  print("Basic I/O properties: failed (test case error!)");
  print("Error states: ", tank:StateSetToString(tank:ErrorStates()));
end

-- input locally free
if faudes.IsInputLocallyFree(tank) then
  print("Input locally free: ok (test case error!)");
else
  print("Input locally free: failed (expected)");
  print("Error states: ", tank:StateSetToString(tank:ErrorStates()));
end

-- fix input locally
faudes.IoFreeInput(tank)
tank:GraphWrite("tmp_ios_tank_ifree.png")
if faudes.IsInputLocallyFree(tank) then
  print("Input locally free: ok (expected)");
else
  print("Input locally free: failed (test case error!)");
end

-- input omega free
if faudes.IsInputOmegaFree(tank) then
  print("Input omega free: ok (expected)");
else
  print("Input omega free: failed (test case error)");
  print("Error states: ", tank:StateSetToString(tank:ErrorStates()));
end

-- remove dummy states
tank2=faudes.IoSystem(tank);
faudes.RemoveIoDummyStates(tank2);
tank2:GraphWrite("tmp_ios_tank_reverted.png")





